pkg/utils/priority.go

```diff
@@ -17,7 +17,7 @@ const SystemCriticalPriority = 2 * int32(1000000000)
 // GetNamespacesFromPodAffinityTerm returns a set of names
 // according to the namespaces indicated in podAffinityTerm.
 // If namespaces is empty it considers the given pod's namespace.
-func GetNamespacesFromPodAffinityTerm(pod *v1.Pod, podAffinityTerm *v1.PodAffinityTerm) sets.Set[string] {
+func getNamespacesFromPodAffinityTerm(pod *v1.Pod, podAffinityTerm *v1.PodAffinityTerm) sets.Set[string] {
```

Comment:
Im a little concerned about this function. I realize this PR does not modify it heavily (modifying public/private), but I'm doing some research on the [namespaceSelector](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#namespace-selector) functionality within pod affinity/anti-affinity and am wondering if we're missing a condition here.

The definition says that:

> You can also select matching namespaces using namespaceSelector, which is a label query over the set of namespaces. The affinity term is applied to namespaces selected by both namespaceSelector and the namespaces field. Note that an empty namespaceSelector ({}) matches all namespaces, while a null or empty namespaces list and null namespaceSelector matches the namespace of the Pod where the rule is defined.

It looks to me like if we have _no_ namespaceSelector defined (i.e. namespaceSelector is `nil`), the current comparison behavior is correct.  However, if we have namespaceSelector as `{}` or an actually defined label selector, we could be missing out on namespaces here.

I think that the way this works is that we could have several namespaces here (not real examples)

 ```sh
❯ k get ns namespace-a -o yaml
apiVersion: v1
kind: Namespace
metadata:
  labels:
    kubernetes.io/metadata.name: namespace-a
    owner: team-a
  name: namespace-a
  
❯ k get ns namespace-b -o yaml
apiVersion: v1
kind: Namespace
metadata:
  labels:
    kubernetes.io/metadata.name: namespace-b
    owner: team-b
  name: namespace-b

❯ k get ns dev-o yaml
apiVersion: v1
kind: Namespace
metadata:
  labels:
    kubernetes.io/metadata.name: dev
    owner: team-b
  name: dev
```

And we could have a pod with a pod antiAffinity:

```yaml
spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - namespaceSelector:
              matchLabels:
                team: team-b
            labelSelector:
              matchExpressions:
              - key: name
                operator: In
                values:
                - tmp
            topologyKey: "kubernetes.io/hostname"
```

I realize this is a somewhat contrived example, but in this case let's assume that this was deployed to `dev`.  It would totally miss out that there may be some pods to check in `namespace-b`, as that would be matched by the namespaceSelector but not by the fact that the `namespace` list was empty.

I'm not sure if this is a blocker or if it just something that we should address at _some point_, I'll leave that up to the maintainers.  I'm not sure how widespread the usage of namespace selectors is (we don't use them), but this may result in surprising behavior.

Reply:
> If we have an actually defined label namespace selector, it would indeed not evaluate namespaces that aren't included in the selector, but this seems to me like the expected behaviour, no? 🤔 I may be misunderstanding something.

What I'm saying here is that if we had a `namespaceSelector` defined, as is allowed by the pod (anti)affinity spec, we're ignoring it in our list of namespaces to check for descheduler, which would _not_ be expected behavior.

The easier case in my mind is imagining we have _no_ `namespaces` defined in the pod affinity term, but we do have a label selector defined in `namespacesSelector`.  In this case, the function `GetNamespacesFromPodAffinityTerm` will return only the pod itself's Namespace, and not any namespaces that match the `namespaceSelector` `labelSelector`.  The kube-scheduler, however, _will_ match the namespaces in the `namespaceSelector`, and will cause differing behavior between the two.

My question is: does this matter enough to fix now?  This PR isn't _introducing_ this behavior, it's pre-existing.  Perhaps this could be just follow-up work for the `nodeFit` redesign.

Reply:
Gotcha, I see what you mean - `getNamespacesFromPodAffinityTerm` does not take into account the `namespaceSelector`. Good catch! This does seem like a missing feature. The PR is already rather large and patches one thing but this could definitely cause issues for folks. It would be worth opening a separate issue so that it's tracked and can be tackled in a follow-up PR.

Reply:
I think that you're right.  I've created this issue:

https://github.com/kubernetes-sigs/descheduler/issues/1361

and resolved my comments, I'll approve and we can follow up.  @knelasevero @a7i - ok to slide this issue into consideration for the nodeFit re-design?

Reply:
Perfect, thank you @logyball! 😄 

Resolving this for now - we can continue the discussion about `namespaceSelector` in the follow-up issue that you created 👍